From c5389bb584f3f4a0fb4801f03277a22a374acfa3 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:29:39 -0400
Subject: [net] openvswitch: Fix race

Message-id: <8848028e0bad7db79abb8e2f4ea0cc9386ad2528.1410531783.git.jbenc@redhat.com>
Patchwork-id: 93993
O-Subject: [RHEL7.1 net PATCH 06/50] openvswitch: Fix race.
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit 42ee19e2939277a5277c307e517ce2d7ba5f0703
Author: Jarno Rajahalme <jrajahalme@nicira.com>
Date:   Sat Feb 15 17:42:29 2014 -0800

    openvswitch: Fix race.

    ovs_vport_cmd_dump() did rcu_read_lock() only after getting the
    datapath, which could have been deleted in between.  Resolved by
    taking rcu_read_lock() before the get_dp() call.

    Signed-off-by: Jarno Rajahalme <jrajahalme@nicira.com>
    Signed-off-by: Pravin B Shelar <pshelar@nicira.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/datapath.c b/net/openvswitch/datapath.c
index 862a42b..ace4695 100644
--- a/net/openvswitch/datapath.c
+++ b/net/openvswitch/datapath.c
@@ -1762,11 +1762,12 @@ static int ovs_vport_cmd_dump(struct sk_buff *skb, struct netlink_callback *cb)
  int bucket = cb->args[0], skip = cb->args[1];
  int i, j = 0;
 
+ rcu_read_lock();
  dp = get_dp(sock_net(skb->sk), ovs_header->dp_ifindex);
- if (!dp)
+ if (!dp) {
+  rcu_read_unlock();
   return -ENODEV;
-
- rcu_read_lock();
+ }
  for (i = bucket; i < DP_VPORT_HASH_BUCKETS; i++) {
   struct vport *vport;
 
-- 
1.7.1